cpu/stm32/cpu_common.c

```diff
@@ -51,343 +77,350 @@ static const uint8_t apbmul[] = {
     }
 }
 
-#if defined(CPU_FAM_STM32L4) || defined(CPU_FAM_STM32G4) || \
-    defined(CPU_FAM_STM32L5) || defined(CPU_FAM_STM32U5)
-void periph_lpclk_en(bus_t bus, uint32_t mask)
+static void _rcc_reg_clr(volatile uint32_t *reg, uint32_t mask)
 {
-    switch (bus) {
-        case APB1:
-            RCC->APB1SMENR1 |= mask;
-            break;
-        case APB12:
-            RCC->APB1SMENR2 |= mask;
-            break;
-        case APB2:
-            RCC->APB2SMENR |= mask;
-            break;
-        case AHB1:
-            RCC->AHB1SMENR |= mask;
-            break;
-        case AHB2:
-#if defined(CPU_FAM_STM32U5)
-            RCC->AHB2SMENR1 |= mask;
-#else
-            RCC->AHB2SMENR |= mask;
-#endif
-            break;
-#if defined(CPU_FAM_STM32U5)
-        case AHB22:
-            RCC->AHB2SMENR2 |= mask;
-            break;
-#endif
-        case AHB3:
-            RCC->AHB3SMENR |= mask;
-            break;
-        default:
-            DEBUG("unsupported bus %d\n", (int)bus);
-            break;
+    assert(reg);
+
+    if (IS_ACTIVE(RCC_REG_IS_ATOMIC)) {
+        *reg = mask;
+    }
+    else {
+        const int irq_state = irq_disable();
+        *reg &= ~(mask);
+        irq_restore(irq_state);
     }
 }
 
-void periph_lpclk_dis(bus_t bus, uint32_t mask)
+uint32_t periph_apb_clk(bus_t bus)
 {
-    switch (bus) {
-        case APB1:
-            RCC->APB1SMENR1 &= ~(mask);
-            break;
-        case APB12:
-            RCC->APB1SMENR2 &= ~(mask);
-            break;
-        case APB2:
-            RCC->APB2SMENR &= ~(mask);
-            break;
-        case AHB1:
-            RCC->AHB1SMENR &= ~(mask);
-            break;
-        case AHB2:
-#if defined(CPU_FAM_STM32U5)
-            RCC->AHB2SMENR1 &= ~(mask);
-#else
-            RCC->AHB2SMENR &= ~(mask);
-#endif
-            break;
-#if defined(CPU_FAM_STM32U5)
-        case AHB22:
-            RCC->AHB2SMENR2 &= ~(mask);
-            break;
-#endif
-        case AHB3:
-            RCC->AHB3SMENR &= ~(mask);
-            break;
-        default:
-            DEBUG("unsupported bus %d\n", (int)bus);
-            break;
+#ifdef CLOCK_APB2
+    if (bus == APB2) {
+        return CLOCK_APB2;
     }
+#else
+    (void)bus;
+#endif
+    return CLOCK_APB1;
+}
+
+uint32_t periph_timer_clk(bus_t bus)
+{
+    return periph_apb_clk(bus) * apbmul[bus];
+}
+
+void periph_clk_en(bus_t bus, uint32_t mask)
+{
+    assert(bus < BUS_NUMOF);
+
+   _rcc_reg_set(_rcc_en_reg(bus), mask);
+
+    /* stm32xx-errata: Delay after a RCC peripheral clock enable */
+    __DSB();
 }
-#elif defined(CPU_FAM_STM32F2) || \
-      defined(CPU_FAM_STM32F4) || \
-      defined(CPU_FAM_STM32F7)
+
+void periph_clk_dis(bus_t bus, uint32_t mask)
+{
+    assert(bus < BUS_NUMOF);
+    _rcc_reg_clr(_rcc_dis_reg(bus), mask);
+}
+
+static inline
+void _periph_lpclk_en(bus_t bus, uint32_t mask)
+{
+    assert(bus < BUS_NUMOF);
+    _rcc_reg_set(_rcc_lp_en_reg(bus), mask);
+}
+
+static inline
```

Comment:
```suggestion
MAYBE_UNUSED static inline
```
looks like you need this for Clang

Reply:
Couldn't you just add the same #ifdef HAS_LP_MODE preprocessor macros around the _periph_lpclk_en/dis functions as are used a little bit further below around the periph_lpclk_en/dis functions? They are the only functions that use these functions and nothing from the outside can use them, so when the periph_lpclk_en/dis is not defined, the _periph_lpclk_en/dis functions do not have to be defined either :thinking:

Reply:
Originally durring development, the body of the inline functions were non-trivial, and the goal was to not hide them from the compiler when not used to avoid stale code. I will go the route of adding the `MAYBE_UNUSED` in the event that they ever need extended since it is already setup that way.

Reply:
done